core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java

```diff
@@ -260,16 +263,27 @@ public BucketNodeMap getBucketNodeMap(Session session, PartitioningHandle partit
      */
     private int getDefaultBucketCount(Session session, PartitioningHandle partitioningHandle)
     {
-        // The goal is to have enough buckets to evenly distribute them across tasks created.  The simplest way to do that is to have exactly as many buckets as there are tasks;
-        // then each task gets one bucket.  So let's return the total number of tasks here.  This number, unfortunately, is hard to know precisely because of the variety of ways in
-        // which schedulers create tasks.  We therefore have to estimate.
+        // The default bucket count is used by both remote and local exchanges to assign buckets to nodes and drivers. The goal is to have enough
+        // buckets to evenly distribute them across tasks or drivers. If number of buckets is too low, then some tasks or drivers will be idle.
+        // Excessive number of buckets would make bucket lists or arrays too large.
+
+        // For remote exchanges bucket count can be assumed to be equal to node count multiplied by some constant.
+        int remoteBucketCount;
         if (getRetryPolicy(session) != RetryPolicy.TASK) {
             // Pipeline schedulers typically create as many tasks as there are nodes.
-            return getNodeCount(session, partitioningHandle);
+            remoteBucketCount = getNodeCount(session, partitioningHandle) * 3;
+        }
+        else {
+            // The FTE scheduler usually creates as many tasks as there are partitions.
+            remoteBucketCount = getFaultTolerantExecutionMaxPartitionCount(session) * 3;
         }
-        // The FTE scheduler usually creates as many tasks as there are partitions.
+
+        // For local exchanges we need to multiply MAX_WRITER_COUNT by SCALE_WRITERS_MAX_PARTITIONS_PER_WRITER to account
+        // for local exchanges that are used to distributed data between table writer drivers.
+        int localBucketCount = MAX_WRITER_COUNT * SCALE_WRITERS_MAX_PARTITIONS_PER_WRITER;
+
         // TODO: get the actual number of partitions if PartitioningHandle ever offers it or if we get the partitioning scheme here as a parameter.
-        return getFaultTolerantExecutionMaxPartitionCount(session);
+        return max(remoteBucketCount, localBucketCount);
```

Comment:
I don't understand how this avoids the too-many-buckets scenario.  Should we separate the counts for local and remote exchanges?

Reply:
See https://github.com/trinodb/trino/pull/22000#discussion_r1603663883. This prevents from too little buckets in case of local exchanges